internal/clients/team4/president.go

```diff
@@ -0,0 +1,351 @@
+			ActionTaken: false,
+		}
+	}
+
+	//Separate and sort allocations
+	criticalRequests := make(map[shared.ClientID]shared.Resources)
+	for islandID, request := range resourceRequest {
+		if p.parent.ServerReadHandle.GetGameState().ClientLifeStatuses[islandID] == shared.Critical {
+			criticalRequests[islandID] = request
+		}
+	}
+	sortedCriticalIslands := rankByAllocationSize(criticalRequests)
+
+	nonCriticalRequests := make(map[shared.ClientID]shared.Resources)
+	for islandID, request := range resourceRequest {
+		if p.parent.ServerReadHandle.GetGameState().ClientLifeStatuses[islandID] == shared.Alive {
+			nonCriticalRequests[islandID] = request
+		}
+	}
+	sortedNonCriticalRequests := rankByAllocationSize(nonCriticalRequests)
+
+	finalAllocation := make(map[shared.ClientID]shared.Resources)
+
+	//Limit resources to be distributed from the CP in a linear manner
+	disasterPrediction := p.parent.obs.iifoObs.finalDisasterPrediction
+	resourceThreshold := shared.Resources(0)
+	if p.parent.ServerReadHandle.GetGameState().Season != 1 && disasterPrediction.Confidence >= 0.5 {
+		if disasterPrediction.TimeLeft != 0 {
+			resourceThreshold = shared.Resources(
+				float64(p.parent.ServerReadHandle.GetGameState().Turn) *
+					disasterPrediction.Magnitude /
+					(float64(disasterPrediction.TimeLeft) + float64(p.parent.ServerReadHandle.GetGameState().Turn)))
+		} else if disasterPrediction.TimeLeft == 0 {
+			resourceThreshold = shared.Resources(disasterPrediction.Magnitude)
+		}
+	}
+	remainingResources := p.parent.ServerReadHandle.GetGameState().CommonPool - resourceThreshold
+
+	//Allocations for critical islands
+	for i := 0; i < len(sortedCriticalIslands); i++ {
+		limitedAllocation := p.allocationLimiter(sortedCriticalIslands[i])
+		if remainingResources-limitedAllocation >= 0 {
+			finalAllocation[sortedCriticalIslands[i].Key] = limitedAllocation
+			remainingResources -= limitedAllocation
+		} else {
+			finalAllocation[sortedCriticalIslands[i].Key] = shared.Resources(0)
+		}
+	}
+
+	//Allocations for non-critical islands
+	for i := 0; i < len(sortedNonCriticalRequests); i++ {
+		limitedAllocation := p.allocationLimiter(sortedNonCriticalRequests[i])
+		if remainingResources-limitedAllocation >= 0 {
+			finalAllocation[sortedNonCriticalRequests[i].Key] = limitedAllocation
+			remainingResources -= limitedAllocation
+		} else {
+			finalAllocation[sortedNonCriticalRequests[i].Key] = shared.Resources(0)
+		}
+	}
+
+	return shared.PresidentReturnContent{
+		ContentType: shared.PresidentAllocation,
+		ResourceMap: finalAllocation,
+		ActionTaken: true,
+	}
+}
+
+func (p *president) allocationLimiter(request IslandRequestPair) shared.Resources {
+	ret := shared.Resources(0)
+	if request.Value*5 >= 5*p.parent.ServerReadHandle.GetGameConfig().CostOfLiving {
+		ret = 5 * p.parent.ServerReadHandle.GetGameConfig().CostOfLiving
+	}
+	//weigh by trust
+	ret = shared.Resources(float64(ret) * 2 * p.parent.trustMatrix.GetClientTrust(request.Key))
```

Comment:
Minor thing, just cast p.parent.trustMatrix.GetClientTrust(request.Key) to shared.Resources